documentation/writing_tests.md

```diff
@@ -0,0 +1,1496 @@
+    - [Using delays](#using-delays)
+    - [Exceptions and edge cases](#exceptions-and-edge-cases)
+    - [How tests should look](#how-tests-should-look)
+- [Getting feedback](#getting-feedback)
+
+
+
+## Engineering setup
+
+The Azure SDK tests not only help us verify that our code is correct, they are also a meaningful way to share how to use our code with our customers. Besides being able to be executed by any developer in the world, they will be triggered by automatic systems that help us verify that we haven't made breaking changes, and also to check that the services we're targeting don't show unexpected behaviors, all of which help us have a better level of confidence before releasing anything to the public.
+
+For our Engineering Systems to pick up our tests appropriately, our packages must be configured according to their guidelines. In this section we will go through some of these concepts, and we will provide links that expand them in detail. We will be covering:
+
+- [Engineering goals](#engineering-goals).
+- [CI and nightly test configuration](#ci-and-nightly-test-configuration).
+- [Delivering live tests to our users](#delivering-live-tests-to-our-users).
+
+### Engineering goals
+
+Though the tests must target live resources, we should make sure they only do so when necessary. We must keep in mind the following considerations:
+
+- **Tests should not be flaky.** Tests should pass regardless of who's executing them, when they are running, and how many times they run.
+- **Tests should create the resources they are testing**, either through the code of the tests, or through their configuration files.
+- **The resources created for the tests should be unique.** Running the same test in parallel, multiple times, should not break them.
+- **Avoid calling to timed delays** (like `setTimeout`) to assert that a change happened in the live resources.
+
+Keep in mind that it is valid to have manually created resources during the _development_ of the tests. For example, if you're working with any of the KeyVault clients, you may target a KeyVault created in your personal Azure account. Each one of the individual clients can then assume that a KeyVault provided by environment variables exists, and only focus on creating and managing the resources they're scoped to work with: Keys, Secrets and Certificates. **Ask your team to see if there's a resource already in place for test development.**
+
+Any resource that is not created by the tests must be defined in an [ARM template](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/overview), so that anyone can build a copy of them. This ARM template will be used by the CI pipelines during builds. Under the context of the previous KeyVault example, to automate the creation of the KeyVault that is passed into the tests, it needs to have been defined specifically in the [CI and nightly test configuration](#ci-and-nightly-test-configuration) files. 
```

Comment:
> This ARM template will be used by the CI pipelines during builds. 

"This ARM template will be used by the test pipelines for live test runs. "

Live test runs happen nightly or when someone manually queues them.

Reply:
Thank you!